kvstore/local_kv_store.go

```diff
@@ -0,0 +1,512 @@
+			// Wrap encryption error
+			return nil, fmt.Errorf("for key %s: %w", key, err)
+		}
+		value = string(decryptResult.Plaintext)
+	}
+
+	// Collect all outpoints for the key
+	outpoints := make([]string, len(outputsResult.Outputs))
+	for i, output := range outputsResult.Outputs {
+		outpoints[i] = output.Outpoint
+	}
+
+	return &lookupValueResult{
+		value:       value,
+		outpoints:   outpoints,
+		inputBeef:   outputsResult.BEEF, // Return the original BEEF received
+		lor:         outputsResult,
+		valueExists: true,
+	}, nil
+}
+
+// getOutputs lists all outputs with the specified key in the context
+func (kv *LocalKVStore) getOutputs(ctx context.Context, key string, limit int) (*wallet.ListOutputsResult, error) {
+	// List outputs matching the basket and key tag
+	listArgs := wallet.ListOutputsArgs{
+		Basket:  kv.context,
+		Tags:    []string{key},
+		Include: "entire transactions",
+		Limit:   uint32(limit),
+	}
+	result, err := kv.wallet.ListOutputs(ctx, listArgs, kv.originator)
+	if err != nil {
+		// Return wrapped wallet error
+		return nil, fmt.Errorf("error ListOutputs for key %s: %w", key, err)
+	}
+	return result, nil
+}
+
+// Get retrieves a value for the given key, or returns the defaultValue if not found.
+func (kv *LocalKVStore) Get(ctx context.Context, key string, defaultValue string) (string, error) {
+	if key == "" {
+		return "", ErrInvalidKey
+	}
+	result, err := kv.lookupValue(ctx, key, defaultValue, 5)
+	if err != nil {
+		// If lookup failed, return the error directly.
+		// The lookupValue function is responsible for wrapping specific errors.
+		// The caller can use errors.Is or errors.As to check the type.
+		return defaultValue, err
+	}
+
+	// If lookupValue succeeded but value doesn't exist (e.g., empty outputs),
+	// return the default value without error.
+	if !result.valueExists {
+		// Here, we could potentially return kvstore.ErrKeyNotFound if defaultValue was empty
+		// to distinguish between an empty stored value and a non-existent key.
+		// For now, matching TS behavior: return default, no error.
+		return defaultValue, nil
+	}
+
+	return result.value, nil
+}
+
+// Set stores a value with the given key, returning the outpoint of the transaction output.
+func (kv *LocalKVStore) Set(ctx context.Context, key string, value string) (string, error) {
+	if key == "" {
+		return "", ErrInvalidKey
+	}
+	if value == "" {
+		return "", ErrInvalidValue
+	}
+	lookupResult, err := kv.lookupValue(ctx, key, "", 10)
+	// Handle specific errors from lookup
+	if err != nil && !errors.Is(err, ErrKeyNotFound) { // Proceed if key simply not found, otherwise log/handle error
+		if errors.Is(err, ErrCorruptedState) {
+			// Don't proceed if the state is known to be corrupted?
+			return "", fmt.Errorf("cannot set key %s due to existing corrupted state: %w", key, err)
+		}
+		// Log other lookup errors but attempt to proceed with Set (overwrite)
+		fmt.Printf("Warning: lookupValue failed during Set for key %s: %v\n", key, err)
```

Comment:
it would be better to accept and use slog.Logger instead, so the adopter could choose how he want to have this message logged.